libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/impl/MapLibreWayPointDrawer.java

```diff
@@ -0,0 +1,185 @@
+        TypedArray typedArray = null;
+        try {
+            typedArray = context.obtainStyledAttributes(styleResId, R.styleable.NavigationMapRoute);
+
+            int originWaypointIcon = typedArray.getResourceId(
+                    R.styleable.NavigationMapRoute_originWaypointIcon, R.drawable.ic_route_origin);
+            int destinationWaypointIcon = typedArray.getResourceId(
+                    R.styleable.NavigationMapRoute_destinationWaypointIcon, R.drawable.ic_route_destination);
+
+            createLayers(
+                    mapStyle,
+                    ContextCompat.getDrawable(context, originWaypointIcon),
+                    ContextCompat.getDrawable(context, destinationWaypointIcon),
+                    belowLayerId
+            );
+        } finally {
+            if (typedArray != null) {
+                typedArray.recycle();
+            }
+        }
+    }
+
+    void createLayers(Style mapStyle, Drawable originIcon, Drawable destinationIcon, String belowLayerId) {
+        SymbolLayer wayPointLayer = routeLayerFactory.createWayPointLayer(
+                mapStyle, originIcon, destinationIcon
+        );
+        MapUtils.addLayerToMap(mapStyle, wayPointLayer, belowLayerId);
+    }
+
+    public void setStyle(Style style) {
+        this.mapStyle = style;
+
+        if (route != null) {
+            drawWayPoints(route.legs());
+        }
+    }
+
+    public void setRoute(DirectionsRoute route) {
+        this.route = route;
+
+        if (route != null) {
+            drawWayPoints(route.legs());
+        }
+    }
+
+    @Override
+    public void setVisibility(boolean isVisible) {
+        if (mapStyle == null || !mapStyle.isFullyLoaded()) {
+            return;
+        }
+
+        Layer wayPointLayer = mapStyle.getLayer(WAYPOINT_LAYER_ID);
+        if (wayPointLayer != null) {
+            wayPointLayer.setProperties(visibility(isVisible ? VISIBLE : NONE));
+        }
+    }
+
+    private void drawWayPoints(List<RouteLeg> legs) {
+        if (!mapStyle.isFullyLoaded()) {
+            // The style is not available anymore. Skip processing.
+            return;
+        }
+
+        final List<Feature> wayPointFeatures = new ArrayList<>();
+        for (RouteLeg leg : legs) {
+            Feature originFeature = createWayPointFeature(leg, 0);
+            if (originFeature != null) {
+                wayPointFeatures.add(originFeature);
+            }
+
+            if (leg.steps() != null) {
+                Feature destinationFeature = createWayPointFeature(leg, leg.steps().size() - 1);
+                if (destinationFeature != null) {
+                    wayPointFeatures.add(destinationFeature);
+                }
+            }
+        }
+
+        getSource(mapStyle).setGeoJson(FeatureCollection.fromFeatures(wayPointFeatures));
+    }
+
+    @Nullable
+    private Feature createWayPointFeature(RouteLeg leg, int index) {
```

Comment:
For rendering related things like this, I personally prefer protected to allow customisation, but since we have an interface, implementers could also overwrite the whole class.

Reply:
I'm on your side, I will have a look and make all logic that make sense to protected 👍